File: tests/r_parity/66_spatial_gmm.R
# StatsPAI spatial GMM (SAR-2SLS / SEM-GMM) parity — Module 66 (R side).
#
# Reads data/66_spatial_gmm.csv, rebuilds the identical row-standardised
# 12x12 rook contiguity from the grid coordinates, and runs the spatialreg
# moment-based references:
#   * SAR 2SLS : spatialreg::stsls(y ~ x1 + x2, W2X = FALSE)  [instruments X, WX]
#   * SEM GMM  : spatialreg::GMerrorsar(y ~ x1 + x2)          [Kelejian-Prucha 1999]
# stsls rows carry the sig2n_k (n-k) SEs; GMerrorsar rows are emitted
# point-only to match the Python side (coefficient-SE conventions differ).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(spatialreg)
  library(spdep)
})

MODULE <- "66_spatial_gmm"

df <- read_csv_strict(MODULE)
n <- nrow(df)

dr <- abs(outer(df$grid_row, df$grid_row, "-"))
dc <- abs(outer(df$grid_col, df$grid_col, "-"))
Wd <- (dr + dc) == 1
storage.mode(Wd) <- "double"
lw <- spdep::mat2listw(Wd, style = "W")

# SAR spatial two-stage least squares. coef() = c(Rho, (Intercept), x1, x2).
s2sls <- spatialreg::stsls(y ~ x1 + x2, data = df, listw = lw, W2X = FALSE)
s_co <- coef(s2sls)
s_se <- sqrt(diag(s2sls$var))

# SEM generalized-moments error model.
gm <- spatialreg::GMerrorsar(y ~ x1 + x2, data = df, listw = lw)
gm_co <- coef(gm)  # (Intercept), x1, x2

rows <- list(
  # --- SAR 2SLS (point + n-k SE) ---
  parity_row(MODULE, "sar_gmm_const",
             estimate = unname(s_co["(Intercept)"]),
             se = unname(s_se["(Intercept)"]), n = n),
  parity_row(MODULE, "sar_gmm_x1",
             estimate = unname(s_co["x1"]), se = unname(s_se["x1"]), n = n),
  parity_row(MODULE, "sar_gmm_x2",
             estimate = unname(s_co["x2"]), se = unname(s_se["x2"]), n = n),
  parity_row(MODULE, "sar_gmm_rho",
             estimate = unname(s_co["Rho"]), se = unname(s_se["Rho"]), n = n),
  # --- SEM GMM (point-only) ---
  parity_row(MODULE, "sem_gmm_const",
             estimate = unname(gm_co["(Intercept)"]), n = n),
  parity_row(MODULE, "sem_gmm_x1",
             estimate = unname(gm_co["x1"]), n = n),
  parity_row(MODULE, "sem_gmm_x2",
             estimate = unname(gm_co["x2"]), n = n),
  parity_row(MODULE, "sem_gmm_lambda",
             estimate = unname(gm$lambda), n = n)
)

write_results(MODULE, rows, extra = list(
  sar_gmm = "spatialreg::stsls(W2X=FALSE)",
  sem_gmm = "spatialreg::GMerrorsar",
  weights = "rook contiguity, row-standardised (style='W')"
))

File: tests/r_parity/66_spatial_gmm.py
"""StatsPAI spatial GMM (SAR-2SLS / SEM-GMM) parity — Module 66 (Python side).

Generates a deterministic 12x12 rook-lattice spatial-error DGP and fits the
two moment-based spatial estimators whose conventions align with
``spatialreg``:

  * ``sp.sar_gmm(..., w_lags=1)`` — Kelejian-Prucha spatial two-stage least
    squares with instruments ``[X, WX]`` and endogenous ``WY``. Matches
    ``spatialreg::stsls(..., W2X = FALSE)`` on both coefficients and the
    ``sig2n_k`` (n-k) standard errors.
  * ``sp.sem_gmm(...)`` — Kelejian-Prucha (1999) generalized-moments spatial
    error model. Matches ``spatialreg::GMerrorsar`` on the coefficients and
    the spatial-error parameter ``lambda``. Emitted point-only: the two
    implementations use different residual-variance estimators for the
    coefficient SEs (~1.5% apart) and sp does not return a lambda SE, so the
    SE rows are not a like-for-like join.

``sp.sarar_gmm`` is deliberately *not* included: its joint feasible-GS lag +
GM-error path does not match ``spatialreg::gstsls`` (different moment
sequence), so no honest parity row exists yet.

Both sides rebuild the identical rook contiguity from the grid coordinates,
row-standardise it (spdep style = "W"), and see byte-identical W.

Tolerance: rel_est < 1e-6 (machine). The spatial 2SLS estimator is a
closed-form projection given the instruments, so it agrees to machine
tolerance; the GM error coefficients likewise land on the same root.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "66_spatial_gmm"
SIDE = 12  # 12 x 12 lattice -> n = 144


def _rook_W(grid_row: np.ndarray, grid_col: np.ndarray) -> np.ndarray:
    """Binary rook contiguity from integer grid coordinates."""
    dr = np.abs(grid_row[:, None] - grid_row[None, :])
    dc = np.abs(grid_col[:, None] - grid_col[None, :])
    return ((dr + dc) == 1).astype(float)


def make_data(seed: int = PARITY_SEED + 1) -> pd.DataFrame:
    """Rook-lattice spatial-error DGP: y = Xb + u, u = (I - 0.5 W)^-1 eps."""
    rng = np.random.default_rng(seed)
    n = SIDE * SIDE
    grid_row = np.repeat(np.arange(SIDE), SIDE)
    grid_col = np.tile(np.arange(SIDE), SIDE)
    W = _rook_W(grid_row, grid_col)
    Wn = W / W.sum(1, keepdims=True)

    x1 = rng.normal(size=n)
    x2 = rng.normal(size=n)
    X = np.column_stack([np.ones(n), x1, x2])
    eps = rng.normal(size=n)
    u = np.linalg.solve(np.eye(n) - 0.5 * Wn, eps)
    y = X @ np.array([1.0, 0.7, -0.4]) + u
    return pd.DataFrame(
        {"y": y, "x1": x1, "x2": x2, "grid_row": grid_row, "grid_col": grid_col}
    )


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    W = _rook_W(df["grid_row"].to_numpy(), df["grid_col"].to_numpy())
    n = len(df)
    rows: list[ParityRecord] = []

    # SAR 2SLS (w_lags=1 == stsls W2X=FALSE): point + sig2n_k SE, bit-exact.
    res_sar = sp.sar_gmm(W, data=df, formula="y ~ x1 + x2", w_lags=1)
    for name in ["const", "x1", "x2", "rho"]:
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"sar_gmm_{name}",
                estimate=float(res_sar.params[name]),
                se=float(res_sar.std_errors[name]),
                n=n,
            )
        )

    # SEM GMM (GMerrorsar): coefficients + lambda bit-exact; point-only
    # because the coefficient-SE variance estimators differ by convention.
    res_sem = sp.sem_gmm(W, data=df, formula="y ~ x1 + x2")
    for name in ["const", "x1", "x2", "lambda"]:
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"sem_gmm_{name}",
                estimate=float(res_sem.params[name]),
                se=None,
                n=n,
            )
        )

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "n": n,
            "lattice": f"{SIDE}x{SIDE} rook contiguity, row-standardised (style='W')",
            "note": (
                "sp.sar_gmm(w_lags=1) vs spatialreg::stsls(W2X=FALSE) "
                "(coefficients + n-k SEs, bit-exact); sp.sem_gmm vs "
                "spatialreg::GMerrorsar (coefficients + lambda, point-only — "
                "SE variance-estimator convention differs)."
            ),
        },
    )


if __name__ == "__main__":
    main()
